docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,350 @@
+SGX Enclave Memory Manager
+=================================
+
+## Motivation ##
+
+An enclave's memory is backed by special reserved region in RAM, called Enclave Page Cache(EPC). Enclave memory management tasks include allocating/reserving virtual address ranges, committing physical EPC pages, changing EPC page permissions or page type, or removing EPC pages. Those tasks require collaboration between trusted runtime, untrusted runtime and the OS. The SGX enclave memory manager (EMM) serves as a central component in the enclave trusted runtime that abstracts the interaction with untrusted runtime for all memory management flows and provides APIs for its clients to reserve virtual address ranges, commit EPC memory to the reserved address ranges, modify attributes of the reserved/committed pages.   
+
+For details of specific memory management related flows, please refer to [the SGX EDMM driver API spec](https://github.com/openenclave/openenclave/pull/3755/files). The public EMM APIs defined here are most likely invoked by some intermediate runtime level components for specific usages, such as dynamic heap/stack, mmap, mprotect, higher level language JIT compiler, etc.   
+
+## User Experience ##
+
+**Allocate, Deallocate Enclave Memory**
+
+The EMM provides an API, sgx_mm_alloc, for clients to request enclave memory allocations. An enclave memory allocation represents both a reserved virtual address range and a commitment of EPC pages.  EPC pages are committed for enclaves via special SGX instructions: loaded by EADD/EEXTEND before EINIT or dynamically added using EAUG followed by EACCEPT. 
+
+The sgx_mm_alloc API allows clients to specify one of three committing modes for an allocation: 
+- EMA_RESERVE, only the virtual address range is reserved. No EPC pages will be committed in this mode.  
+- EMA_COMMIT_NOW: reserves and commits physical EPC committed upon allocation. EACCEPT will be done immediately.
+- EMA_COMMIT_ON_DEMAND: EACCEPT is done on demand, see below on committing and uncommitting.
+
+An allocation, once created, will own its address range until the deallocation API, sgx_mm_dealloc, is called upon. No two active allocations can have overlapping address ranges.
+
+**Commit, Uncommit Enclave Memory**
+
+When a page in COMMIT_ON_DEMAND allocations is accessed, a page fault occurs if the page was not yet commited.  The EMM will perform EACCEPT to commit the EPC page on page fault after OS doing EAUG. 
+
+The clients can also call the EMM commit API, sgx_mm_commit, to proactively commit specific sub-regions in a COMMIT_ON_DEMAND allocation to avoid future page fault.
+
+Some EMM clients, <i>e.g.</i>, a dynamic code loader wishing to load code on page fault, can register a custom handler for page fault at the time of allocation request. In the custom page fault handler, it can invoke an API, sgx_mm_commit_data, to commit and load data to newly committed EPC page at the same time as supported by EACCEPTCOPY. 
```

Comment:
```suggestion
Some EMM clients, <i>e.g.</i>, a dynamic code loader wishing to load code on page faults, can register a custom handler for page faults at the time of allocation request. In the custom page fault handler, it can invoke an API, sgx_mm_commit_data, to commit and load data to newly committed EPC page at the same time as supported by EACCEPTCOPY. 
```